### PR TITLE
Scheduler prefers nodes without connection failures

### PR DESCRIPTION
**Pull request by dongluochen**
Current swarm scheduler may keep trying the same node until its state changes to 'Unhealthy'. This change makes scheduler aware of connection failures and prefer nodes without failures. 

Signed-off-by: Dong Chen <dongluo.chen@docker.com>

### CODE REVIEW

File: cluster/engine.go [hunk 1]
```diff
@@ -209,6 +209,18 @@ func (e *Engine) IsHealthy() bool {
 	return e.state == stateHealthy
 }
 
+// HealthIndicator returns degree of healthiness between 0 and 100.
+// 0 means node is not healthy (unhealthy, pending), 100 means last connectivity was successful
+// other values indicate recent failures but haven't moved engine out of healthy state
+func (e *Engine) HealthIndicator() int {
```
Comment (vieux) [inline cluster/engine.go]: I'm curious, why not int64 here ?

File: cluster/engine.go [hunk 2]
```diff
@@ -209,6 +209,18 @@ func (e *Engine) IsHealthy() bool {
...
+	e.RLock()
+	e.RUnlock()
```
Comment (aaronlehmann) [inline cluster/engine.go]: Is this statement meant to be deferred?

File: cluster/engine.go (hunk 1 above)
Comment (dongluochen) [inline cluster/engine.go]: You are right. It's better to do the conversion here.

File: cluster/engine.go (hunk 2 above)
Comment (dongluochen) [inline cluster/engine.go]: Good catch. Thanks!

**Comment by vieux** [comment]
LGTM

ping @abronan

**Comment by abronan** [comment]
@dongluochen This may not be practical to do it (and we might hit timing issues on the CI), but that would be cool if we had the integration tests testing 3 or more engines with different health scores to test the impact of the health score on the scheduling more thoroughly.

LGTM though this should not block the merge.

**Pull request merged**